### ISSUE TITLE
RFC: real-time gate-model prediction in the daemon

### ISSUE THREAD

**Issue by jm709** [jm709/pscanner#77]
## Motivation

The XGBoost gate model produced by \`pscanner ml train\` is currently an offline artifact — \`scripts/analyze_model.py\` evaluates it against a held-out test split, and that's it. There's no surface for using the model's predictions in live trading decisions.

This RFC proposes the architecture for scoring incoming trades in real time, emitting alerts when the model thinks a trade has positive edge, and routing those alerts through the existing paper-trading infrastructure to track live performance.

## Goals

1. **Score every newly-observed trade** from the daemon's trade collector against the loaded gate model within ~1 second of observation.
2. **Emit a structured alert** when \`pred > min_pred\` (configurable threshold) and \`pred > implied_prob_at_buy\` (the same gate the offline metric uses).
3. **Drive paper trades** via a new \`SignalEvaluator\` so the live system's edge is observable in \`pscanner paper status\` and the per-source breakdown table.
4. **Survive daemon restarts** — wallet/market state shouldn't have to rebuild from \`corpus_trades\` on every start (multi-hour cold start would be unacceptable).
5. **Gate the live signal** behind \`accepted_categories\` (sports + esports per the current model's preprocessor.json) so we don't trade where the offline analysis showed no edge.

## Out of scope (deferred)

- **Live order execution.** Paper trades only. Real money is a separate, much bigger problem (key management, slippage, regulatory).
- **Model retraining cadence.** This RFC assumes the model artifact updates via human-driven \`ml train\` runs; automated retraining is a follow-up.
- **Multi-platform.** Polymarket only. Kalshi/Manifold extensions wait for the per-platform feature-engineering pipelines (Stage 2 of #35).
- **Cross-detector composition.** The gate model is a standalone signal; no AND/OR logic with smart_money, velocity, mispricing, monotone alerts in v1. Each evaluator stays independent.

### Q1: Score every trade vs. score only watched-wallet trades?

The model was trained on ALL trades in \`corpus_trades\`, including from non-watched wallets. But the daemon's \`TradeCollector\` only polls \`/activity\` for **watched** wallets. So we have two paths:

**(a) Watched-wallet trades only (proposed for v1).** Reuses the existing \`TradeCollector.subscribe_new_trade\` callback. Wallet-quality features are populated correctly because we have history for these wallets. Misses general trade flow.

**(b) All trades.** Would require a new ingestion path (Polymarket WS \`/ws/user\` is auth-required and per-wallet; subgraph polling is high-latency; CTF-Exchange WS doesn't exist). Plus, wallet-quality features for unknown wallets default to nulls/zeros, so the model's predictions on those rows are poor anyway (the offline evaluation showed wallet-quality features round to zero gain at 60/20/20, but at 70/15/15 they appear in the top 25 by gain — so unknown wallets are a real drag on prediction quality).

Recommendation: **(a)** for v1. Captures the high-value "watched wallets we already trust" surface. Revisit in v2 if we want broader coverage.

### Q2: Persistent state — snapshot or rebuild?

The streaming history provider's wallet/market state is in-memory today (\`pscanner.corpus.features.StreamingHistoryProvider\`). For a daemon, we need to persist across restarts. Options:

**(a) Periodic snapshot to SQLite.** New tables \`wallet_state_snapshot\` + \`market_state_snapshot\`. Snapshot every N minutes; restore on daemon start. Trade-off: state may be stale by N minutes on restart; tail of trades replayed from \`wallet_trades\`.

**(b) Incremental update tables.** Maintain \`wallet_state_live\` and \`market_state_live\` in SQLite, updated atomically per trade. No replay needed. Higher write amplification but always-current.

**(c) Cold-rebuild from corpus_trades on every start.** Already-implemented logic, but for ~15M trades this is the multi-hour bound from #67/#39's analysis. Unacceptable.

Recommendation: **(b)** — incremental update tables. Each trade callback reads the current state row, applies the fold, writes back. Single-row UPDATEs are cheap. Daemon startup is O(1) instead of O(corpus size).

### Q3: Model artifact location and hot-reload?

The current artifact lives at \`<output_dir>/model.json\` + \`preprocessor.json\`. Daemon needs to know which artifact to load. Options:

**(a) Config-pinned path.** Add \`pscanner.config.GateModelConfig\` with \`artifact_dir: Path\`. Daemon loads on startup; restart required to pick up a new model. Simple.
**(b) Symlink convention.** \`./models/current\` → latest artifact dir. Daemon checks the symlink target on a cadence (every 10 min); reloads if changed. Hot reload without restart.
**(c) File-watcher hot reload.** \`watchdog\` or similar; reload as soon as the artifact dir changes.

Recommendation: **(a)** for v1. Simpler. Hot reload is a v2 nice-to-have.

### Q4: Feedback loop — do we score our own paper trades?

The paper trader books simulated trades into \`paper_trades\`. If those trades' positions/prices feed back into the streaming history provider's wallet state, the gate model would see its own past decisions and could overfit on simulated history that doesn't exist in reality.

Recommendation: **paper trades do NOT update the streaming history provider's state.** Only real on-chain trades observed via \`TradeCollector\` get folded in. The provider's state is the same one used by build-features at training time.

### Q5: Drift detection?

Markets change. Wallet behaviors change. The model trained on data through 2026-04 may be stale by 2026-08. We should track:

- Rolling 7-day realized edge of taken bets vs. test_edge_filtered baseline.
- Per-category edge over time.
- Prediction distribution drift (if the bottom decile suddenly stops getting taken bets, something shifted).

Recommendation: separate observability work. File as a v2 follow-up.

## Architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                          pscanner daemon                         │
│                                                                  │
│   TradeCollector ──[on_trade]──> GateModelDetector              │
│        │                              │                          │
│        │                              ├─► load_model() once      │
│        │                              ├─► persistent_provider    │
│        │                              │      .observe(trade)     │
│        │                              ├─► features = compute…    │
│        │                              ├─► pred = model.predict   │
│        │                              ├─► if pred > threshold:   │
│        │                              │      emit("gate_buy")    │
│        │                              ▼                          │
│        ▼                          AlertSink ──► PaperTrader      │
│   wallet_trades                                     │             │
│        │                                            ├─ GateEvaluator
│        ▼                                            │   parse()  │
│   wallet_state_live  ◄───────── persistent_provider │   gate     │
│   market_state_live                                 │   size     │
│                                                     ▼             │
│                                               paper_trades        │
└─────────────────────────────────────────────────────────────────┘
```

Three new components, each tracked as a separate implementation issue:

1. **Persistent streaming history provider** — adapt \`StreamingHistoryProvider\` to read+write \`wallet_state_live\` and \`market_state_live\` SQLite tables instead of in-memory dicts. Bootstrapping path: cold-start from corpus_trades the first time, then incremental forever.
2. **Gate-model detector** — new detector class subscribing to \`TradeCollector\` callbacks. Loads model on init, computes features per trade via the persistent provider, emits \`"gate_buy"\` alerts when \`pred > min_pred\`.
3. **Gate-model evaluator + paper trading integration** — new \`GateModelEvaluator\` implementing \`SignalEvaluator\`. Adds \`"gate_buy"\` to the \`DetectorName\` literal in \`alerts/models.py\`. Per-source breakdown row in \`paper status\`.

## Definition of done (v1)

- Daemon can be started with a model artifact configured; logs \`gate_model.loaded\` once with model version + accepted_categories.
- For every observed watched-wallet trade, daemon emits \`gate_buy\` alert iff \`pred > min_pred\` AND \`pred > implied_prob_at_buy\` AND \`top_category in accepted_categories\`.
- Paper trader books a paper trade per accepted alert, sized at \`gate_model.position_fraction\` (config-validated).
- \`pscanner paper status\` shows a \`gate_model\` row in the per-source breakdown.
- Daemon restart reloads wallet_state_live + market_state_live in O(1) seconds; recent trades since the last snapshot are replayed if needed.
- Per-trade latency: \`< 100 ms p99\` from trade observation to alert emission (model.predict on a single row is microseconds; the bound is feature compute).

## Test plan

- Unit tests for the persistent provider (snapshot/restore semantics).
- Integration test: synthetic trade → feature compute → predict → alert emit → paper trade entry, end-to-end against \`tmp_db\`.
- Replay test: process 1K synthetic trades through the daemon and the offline build-features pipeline; assert resulting paper trades match what the offline pipeline + post-hoc gate would have produced. This is the parity contract that catches drift in feature engineering.

**Comment by jm709** [comment]
## Update: Q1 resolved — score top-markets-by-volume, not watched-wallets only

Discussed offline. The watched-wallets-only scope was too narrow — it caps the gate-model's surface at ~700 wallets and misses the bulk of trade flow on high-volume markets. The right v1 scope is:

**Subscribe to open markets matching a volume threshold (e.g., \`volume_24h > $1M\`), then score every trade on those markets through the gate model — regardless of which wallet placed the trade.**

This makes #78 (persistent state) and #79 (gate detector) larger:

- **#78 implication:** \`LiveHistoryProvider\` must handle unknown wallets gracefully. New wallets observed on a watched market get an \`empty_wallet_state\` row inserted on first trade; their wallet-quality features start as nulls/zeros and accumulate from that point. The offline model trained on \`OneHotEncoder\` with \`__none__\` token, so null wallet-quality features are a known regime — model falls back to market-state features.

- **#79 implication:** Need a new \`MarketScopedTradeCollector\` (parallel to the existing \`TradeCollector\`). It enumerates open markets passing the volume floor (reusing \`MarketTickCollector\`'s volume-floor query against gamma), polls \`/trades?market=X\` per market at staggered intervals (~30s freshness), and dispatches trades into the existing \`subscribe_new_trade\` callback fan-out. Rate-limit budget: ~50 RPM on data-api, so cap at top-50 markets refreshed every minute, or top-100 every two minutes.

- **#80 implication:** unchanged. The evaluator consumes alerts; the alert source doesn't matter to it.

### New per-trade throughput concern

Top-50 markets at ~1-2 trades/sec each = 50-100 trades/sec scoring rate. Per-trade scoring cost (SQLite state read + feature compute + XGBoost predict + state UPDATE) is ~10 ms; at 100 trades/sec we'd saturate one core. The detector must:

- Schedule scoring as async tasks (don't block the polling loop).
- Pre-screen: skip SELL trades, skip non-binary outcomes, skip trades below a notional floor before paying the scoring cost.
- Bounded queue with drop-on-full and a structured warning log (parallel to \`tick_stream.subscriber_queue_full\`).

Profile during implementation; tune scope (top-N, polling interval) if throughput becomes the bound.

### Updated definition-of-done implications

- \`pscanner.poly.data.DataClient\` (or new) gains \`get_market_trades(condition_id, since_ts) -> list[Trade]\`.
- \`MarketScopedTradeCollector\` discovers and polls top-volume markets; emits trades into the existing trade-callback bus that \`GateModelDetector\` (#79) subscribes to.
- New config section \`gate_model_market_filter\` with \`min_volume_24h_usd\` (default $1M), \`max_markets\` (default 50), \`poll_interval_seconds\` (default 60).

**Comment by jm709** [comment]
## v1 scope further narrowed: esports-only

Discussed offline. The first deployment of the live gate model targets **esports markets only**, then expands to sports + esports as confidence grows. Reasoning:

- **Smallest surface to test the loop end-to-end.** Esports has tens of open markets at any time, not hundreds. Trades-per-second across esports is order-1, not order-100. Removes throughput as a v1 concern.
- **Wallet bootstrap pressure is minimal.** Esports markets see far fewer distinct wallets than sports. Cold-start cache fills fast; ongoing unknown-wallet rate is low.
- **Edge is the highest per-bet** (+10.7% on the v2 model's test split per the run today, vs +6.9% for sports). High signal-to-noise per observation.
- **Lowest blast radius if something goes wrong.** Paper-trade volume is bounded by trade flow; a misconfigured detector can't book thousands of bad trades because the signal source is small.

### Implication for the design

Add a new market-filter config at \`scheduler\` time:

\`\`\`python
class GateModelMarketFilterConfig(_Section):
    enabled: bool = False
    accepted_categories: tuple[str, ...] = ("esports",)   # v1: esports only
    min_volume_24h_usd: float = 100_000   # low floor — esports markets are smaller
    max_markets: int = 50
    poll_interval_seconds: int = 60
\`\`\`

The \`MarketScopedTradeCollector\` (sub-component of #79) enumerates open markets via gamma, filters by \`market.top_category in accepted_categories\` AND \`market.volume_24h_usd >= min_volume_24h_usd\`, then polls trades for each.

The per-trade gate inside \`GateModelDetector\` continues to check \`features.top_category in accepted_categories\` as a safety net — the trade-collector filter is the primary mechanism, but defensive double-check catches any misclassification.

### Expansion path

- **v1.0:** esports only.
- **v1.1:** sports + esports (matches the model's full \`accepted_categories\`). Throughput jumps to top-100 markets, ~50-100 trades/sec.
- **v2:** drop the volume floor, expand to all categories the model says it has edge in. Drift detection becomes mandatory.

The v1.0 → v1.1 transition is a config flag flip; no code change. So we stage and gain confidence with the same architecture.

### Q1 → resolved

Q1 (watched-wallets vs top-markets-by-volume) is resolved. The answer is **top-volume open markets within \`accepted_categories\`**, where \`accepted_categories=("esports",)\` for v1.0 and \`("sports", "esports")\` for v1.1.
